### PR TITLE
[proxy]: parse proxy protocol TLVs with aws/azure support

### PR DESCRIPTION
**Pull request by conradludgate**
AWS/azure private link shares extra information in the "TLV" values of the proxy protocol v2 header. This code doesn't action on it, but it parses it as appropriate.

### CODE REVIEW

File: proxy/src/protocol2.rs [hunk 1]
```diff
@@ -164,22 +193,107 @@ pub(crate) async fn read_proxy_protocol<T: AsyncRead + Unpin>(
...
+            let _dst_addr = addr.get_u32();
+            let src_port = addr.get_u16();
+            let _dst_port = addr.get_u16();
             Some(SocketAddr::from((src_addr, src_port)))
         }
         36 => {
-            let src_addr: [u8; 16] = addresses[0..16].try_into().unwrap();
-            let src_port = u16::from_be_bytes(addresses[32..34].try_into().unwrap());
+            let src_addr = Ipv6Addr::from_bits(addr.get_u128());
+            let _dst_addr = addr.get_u128();
```
Comment (awarus) [inline proxy/src/protocol2.rs:208]: should we use here Ipv6Addr::from_bits() fn?

Comment (conradludgate) [reply proxy/src/protocol2.rs:208]: Could do for consistency, but this data is discarded. The `get_u128` is only there to advance the cursor.

File: proxy/src/protocol2.rs [hunk 2]
```diff
@@ -216,6 +330,25 @@ impl<T: AsyncRead> ChainRW<T> {
...
+    value: Bytes,
+}
+
+fn read_tlv(b: &mut BytesMut) -> Option<Tlv> {
+    if b.len() < 3 {
+        return None;
+    }
+    let kind = b.get_u8();
+    let len = usize::from(b.get_u16());
+    if b.len() < len {
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/protocol2.rs:345]: Hmm, not sure what upper limit we could check?

File: proxy/src/protocol2.rs (hunk 1 above)
Comment (awarus) [reply proxy/src/protocol2.rs:208]: Yes, not critical, just if we would need to work with dst addr

**Review by awarus — APPROVED** [review]

**Pull request merged**